cmd/soroban-cli/src/commands/contract/info/shared.rs

```diff
@@ -84,18 +113,25 @@ pub async fn fetch_wasm(args: &Args) -> Result<Option<Vec<u8>>, Error> {
             .verify_network_passphrase(Some(&network.network_passphrase))
             .await?;
 
-        get_remote_wasm_from_hash(&client, &hash).await?
+        print.globeln(format!(
+            "Downloading contract spec for wasm hash: {wasm_hash}"
+        ));
+        Ok((
+            Some(get_remote_wasm_from_hash(&client, &hash).await?),
+            None,
+            Some(network.clone()),
+        ))
     } else if let Some(contract_id) = &args.contract_id {
         let contract_id =
             contract_id.resolve_contract_id(&args.locator, &network.network_passphrase)?;
+        let contract_address = xdr::ScAddress::Contract(xdr::Hash(contract_id.0)).to_string();
+        print.globeln(format!("Downloading contract spec: {contract_address}"));
         let res = wasm::fetch_from_contract(&contract_id, network).await;
         if let Some(ContractIsStellarAsset) = res.as_ref().err() {
-            return Ok(None);
+            return Ok((None, Some(contract_address), Some(network.clone())));
         }
-        res?
+        Ok((Some(res?), Some(contract_address), Some(network.clone())))
     } else {
-        unreachable!("One of contract location arguments must be passed");
-    };
-
-    Ok(Some(wasm))
+        return Err(Error::MalformedWasmOrWasmHashOrContractId);
```

Comment:
If my suggestion above is executed I believe this error can be removed, since the calling code can confidently handle all the cases because the union will be defined specifically to handle all the cases.